### PR TITLE
Add tutorial on events and triggers

### PR DESCRIPTION
**Pull request by wthrowe**
[Rendered](http://hosting.astro.cornell.edu/~wthrowe/spectre_doc_events_and_triggers/tutorial_events_and_triggers.html)

Short user tutorial on the use of events and triggers.  Probably needs more information on a bunch of things, but it is difficult for me to judge as the code author, so please let me know what to expand on.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by nilsdeppe** [comment]
@knelli2 @yoonso0-0 would you be willing to review this? Feel free to decline :)

**Comment by yoonso0-0** [comment]
I can help :)

File: docs/Tutorials/EventsAndTriggers.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+See LICENSE.txt for details.
+\endcond
+
+# Events and triggers {#tutorial_events_and_triggers}
+
+The "events and triggers" infrastructure (and the related "events and
+dense triggers") provide some control of SpECTRE executable execution
+from the input file contents.  It does not generally have major
+effects on the flow of the algorithm, but is primarily used for
+observations and minor adjustments.
```
Comment (yoonso0-0) [inline docs/Tutorials/EventsAndTriggers.md]: I think it'd be good to clarify the meaning of the term 'observation' which we use here and after since this is one of the tutorial docs. Some phrases like 'writing/recording simulation data' would be good.

File: docs/Tutorials/EventsAndTriggers.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+Events::ChangeSlabSize "ChangeSlabSize".)  In most cases, however, their
+effects on program execution are ephemeral, causing some data to be
+written to a file or standard output but not changing anything about
+the simulation state.
+
+### What is a trigger?
+
+A \ref Trigger "trigger" is an input-file-creatable object that
+controls when events are run.  They are checked periodically, once per
+slab in an evolution executable, and once per iteration in an elliptic
```
Comment (yoonso0-0) [inline docs/Tutorials/EventsAndTriggers.md]: \ref Slab may help a bit

File: docs/Tutorials/EventsAndTriggers.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+solve.  At each check, a trigger may fire, causing its associated
+events to run, or not.  Triggers must always be written to be
+independent of element-specific state such as the local values of the
+system variables, so that a trigger will give a consistent result over
+the entire domain at each check.
+
+### What is a dense trigger?
+
+A \ref DenseTrigger "dense trigger" is similar to a trigger, but uses
+time-stepper dense output to provide more precise control of the time
```
Comment (yoonso0-0) [inline docs/Tutorials/EventsAndTriggers.md]: Q. What is the time-stepper dense output?

File: tests/InputFiles/ScalarWave/PlaneWave1DEventsAndTriggersExample.yaml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    UpwindPenalty:
+  DiscontinuousGalerkin:
+    Formulation: StrongInertial
+    Quadrature: GaussLobatto
+
+EventsAndDenseTriggers:
+
+EventsAndTriggers:
+
+# [multiple_events]
```
Comment (yoonso0-0) [inline tests/InputFiles/ScalarWave/PlaneWave1DEventsAndTriggersExample.yaml:52]: It'd be more clear if the line `EventsAndTriggers:` is included as well.

File: docs/Tutorials/EventsAndTriggers.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+The `?:` syntax defines an expanded key-value pair, which allows the
+key to be a complicated type, unlike the common `Key: Value` syntax.
+The `-` is the expanded form of a list, which allows entries to have
+complicated types.  Frequently only one event is associated with a
+trigger, so there will only be a single `-` after the `:`, but
+additional events introduced by more `-` characters can appear
+afterwards:
+
+\snippet PlaneWave1DEventsAndTriggersExample.yaml multiple_events
+
```
Comment (yoonso0-0) [inline docs/Tutorials/EventsAndTriggers.md:67]: It would help a lot if there is a description on this snippet based on the contents previously introduced -- which one stands for an event, which one stands for a trigger, and which lines are options for an event or a trigger, etc.

File: docs/Tutorials/EventsAndTriggers.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+The "events and triggers" infrastructure (and the related "events and
+dense triggers") provide some control of SpECTRE executable execution
+from the input file contents.  It does not generally have major
+effects on the flow of the algorithm, but is primarily used for
+observations and minor adjustments.
+
+### What is an event?
+
+An \ref Event "event" is an input-file-creatable object that performs
+some task, frequently an observation.  With the exception of the \ref
```
Comment (yoonso0-0) [inline docs/Tutorials/EventsAndTriggers.md]: Can you provide some concrete examples of event in casual words? Like 'write data to a file',  'record error norms', or 'exit program'. (I cannot think of other good examples for now)

**Review by yoonso0-0 — COMMENTED** [review]
Thanks for writing this up, this is concise and clear . Here I listed some points that might be helpful for new users.

File: docs/Tutorials/EventsAndTriggers.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+Distributed under the MIT License.
+See LICENSE.txt for details.
+\endcond
+
+# Events and triggers {#tutorial_events_and_triggers}
+
+The "events and triggers" infrastructure (and the related "events and
+dense triggers") provide some control of SpECTRE executable execution
+from the input file contents.  It does not generally have major
+effects on the flow of the algorithm, but is primarily used for
```
Comment (knelli2) [inline docs/Tutorials/EventsAndTriggers.md]: \ref dev_guide_parallelization_core_algorithm algorithm

File: docs/Tutorials/EventsAndTriggers.md (hunk 1 above)
Comment (knelli2) [reply docs/Tutorials/EventsAndTriggers.md]: I think a \ref would be good here to the base for an observer (if there is one).

File: docs/Tutorials/EventsAndTriggers.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+observations and minor adjustments.
+
+### What is an event?
+
+An \ref Event "event" is an input-file-creatable object that performs
+some task, frequently an observation.  With the exception of the \ref
+Events::Completion "Completion" event, they cannot have direct effects
+on the algorithm, although they can have indirect effects through
+sending messages.  (An example of such an event is \ref
+Events::ChangeSlabSize "ChangeSlabSize".)  In most cases, however, their
```
Comment (knelli2) [inline docs/Tutorials/EventsAndTriggers.md]: This was confusing upon a first read. If I understood this correctly, clarify that by "cannot have direct effects on the algorithm" you mean that essentially events can't add actions to the pdal which is specified in the metavariables. But you can (potentially) change things in the databox or cache (like the slab size).

File: docs/Tutorials/EventsAndTriggers.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+### What is a trigger?
+
+A \ref Trigger "trigger" is an input-file-creatable object that
+controls when events are run.  They are checked periodically, once per
+slab in an evolution executable, and once per iteration in an elliptic
+solve.  At each check, a trigger may fire, causing its associated
+events to run, or not.  Triggers must always be written to be
+independent of element-specific state such as the local values of the
+system variables, so that a trigger will give a consistent result over
+the entire domain at each check.
```
Comment (knelli2) [inline docs/Tutorials/EventsAndTriggers.md]: This sentence was a bit wordy and confusing, specifically the phrasing of "independent of element-specific state".

File: docs/Tutorials/EventsAndTriggers.md (hunk 3 above)
Comment (knelli2) [reply docs/Tutorials/EventsAndTriggers.md]: A \ref would be helpful here as well.

File: docs/Tutorials/EventsAndTriggers.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+# Events and triggers {#tutorial_events_and_triggers}
+
+The "events and triggers" infrastructure (and the related "events and
+dense triggers") provide some control of SpECTRE executable execution
+from the input file contents.  It does not generally have major
+effects on the flow of the algorithm, but is primarily used for
+observations and minor adjustments.
+
+### What is an event?
```
Comment (knelli2) [inline docs/Tutorials/EventsAndTriggers.md]: I'm personally not a fan of the "question and answer" type of documentation. Also I don't think it's used anywhere else in spectre (if it is used, it's sparse). For consistency I think we should keep it as
- Event
- Trigger
- Dense Trigger
- Input File Syntax

File: docs/Tutorials/EventsAndTriggers.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+If an event and trigger both took no options, the syntax
+
+\snippet PlaneWave1DEventsAndTriggersExample.yaml no_arguments
+
+could be written in the more familiar form
+
+\snippet PlaneWave1DEventsAndTriggersExample.yaml no_arguments_single_line
+
+but we recommend always using the expanded form for consistency.
```
Comment (knelli2) [inline docs/Tutorials/EventsAndTriggers.md:80]: If you don't want users to use this format for triggers/events with no options, then don't put it in. It'll only confuse them more. What about this for this part:
```suggestion
If an event and trigger both took no options, the recommended syntax is still

\snippet PlaneWave1DEventsAndTriggersExample.yaml no_arguments
```

**Review by knelli2 — CHANGES_REQUESTED** [review]
I personally like having as many (useful) links as possible in documentation because learning code is a very non-linear process. When learning about events and triggers, I definitely didn't have a good grasp on the parallelization infrastructure. So rather than trying to look up "algorithm" myself, I find it very useful having these large topics embedded in *all* tutorials, regardless of the when someone "should" know something which is why I have several \ref comments. Something to think about for the future, but I wanted to address it here as well.

Currently this is set up to be a user tutorial, which I think is good. However, I think we also need a dev guide tutorial for this as well because I would have several other questions about events and triggers if I wanted to actually write one or the other, such as:
- What is the general structure for writing a new event/trigger? Are there protocols?
- How do I add them in executables? What is factory creation?
- Do all events/triggers need to be constructable by options?
- Where are events/triggers actually run? On the elements? Their own chare?

However, just for *using* existing events and triggers in input files and existing executables, I think this is good.

File: docs/Tutorials/EventsAndTriggers.md (hunk 1 above)
Comment (wthrowe) [reply docs/Tutorials/EventsAndTriggers.md]: I don't think we have anything appropriate to point to here.  There's a tutorial on visualization, but that's mostly focused on things like ParaView, and we have a dev guide on the really low-level observation code, but you probably don't even need to understand that if you are writing an observer.  (I have added some words to the text here.)

File: docs/Tutorials/EventsAndTriggers.md (hunk 6 above)
Comment (wthrowe) [reply docs/Tutorials/EventsAndTriggers.md]: This is sort of what I was going for with the rest of the paragraph, but it is not ordered in the way that makes sense to think about the concepts, so I'll redo that.

File: docs/Tutorials/EventsAndTriggers.md (hunk 8 above)
Comment (wthrowe) [reply docs/Tutorials/EventsAndTriggers.md]: Kind of the opposite, actually.  They can't change the DataBox.  I guess they can technically do cache updates, but doing those from an array element seems generally unlikely to work.  They cannot do anything except send messages.  Hopefully the reword here helps.

File: docs/Tutorials/EventsAndTriggers.md (hunk 9 above)
Comment (wthrowe) [reply docs/Tutorials/EventsAndTriggers.md]: Reversed the clauses.  Any better?

File: docs/Tutorials/EventsAndTriggers.md (hunk 3 above)
Comment (wthrowe) [reply docs/Tutorials/EventsAndTriggers.md]: Added "(i.e., interpolation or extrapolation)", but I don't think we have any documentation on this topic.

File: docs/Tutorials/EventsAndTriggers.md (hunk 11 above)
Comment (wthrowe) [reply docs/Tutorials/EventsAndTriggers.md:80]: I was trying to teach something beyond how to cut and paste.  If you think it is really a bad idea I can leave the whole thing out, but I don't see any reason to mention this case if we're not using to to explain something.

File: docs/Tutorials/EventsAndTriggers.md (hunk 5 above)
Comment (wthrowe) [reply docs/Tutorials/EventsAndTriggers.md:67]: Done, but unfortunately I can't link one of the events because of doxygen limitations.

**Review by wthrowe — COMMENTED** [review]
> * What is the general structure for writing a new event/trigger? Are there protocols?

As with most things, copy the structure of an existing one.  (They also have abstract base classes.)

> * How do I add them in executables? What is factory creation?

Added a link.

> * Do all events/triggers need to be constructable by options?

That really belongs in the option parsing documentation, not here.  That documentation is, admittedly, out-of-date on the subject.

> * Where are events/triggers actually run? On the elements? Their own chare?

Added some text.

Edit: Updated rendered version in original post.

File: docs/Tutorials/EventsAndTriggers.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+elements.  It does not generally have major effects on the flow of the
+\ref dev_guide_parallelization_core_algorithm "algorithm", but is
+primarily used for observations, such as writing volume data to H5
+files, and minor simulation adjustments.
+
+### Events
+
+An \ref Event "event" is an input-file-creatable object that performs
+some task that (with the exception of the \ref Events::Completion
+"Completion" event), does not directly affect the execution of the
```
Comment (knelli2) [inline docs/Tutorials/EventsAndTriggers.md]: [nitpick] No comma after the parentheses.

File: docs/Tutorials/EventsAndTriggers.md [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+algorithm on the element.  The effects of events are limited to
+sending messages.  The most commonly used events (such as the \ref
+dg::Events::ObserveErrorNorms "ObserveErrorNorms" event) send data to
+be written to disk, but have no long-term effects on the simulation
+state.  Others, (such as \ref Events::ChangeSlabSize "ChangeSlabSize")
```
Comment (knelli2) [inline docs/Tutorials/EventsAndTriggers.md]: [nitpck] No comma

File: docs/Tutorials/EventsAndTriggers.md (hunk 11 above)
Comment (knelli2) [reply docs/Tutorials/EventsAndTriggers.md:80]: This was more of a nitpick so I think it's ok as is.

File: docs/Tutorials/EventsAndTriggers.md [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+can have indirect effects when \ref dev_guide_parallelization_actions
+"actions" react to the messages they send.
+
+### Triggers
+
+A \ref Trigger "trigger" is an input-file-creatable object that
+controls when events are run.  They are checked periodically, once per
+Slab in an evolution executable, and once per iteration in an elliptic
+solve.  At each check, a trigger may fire, causing its associated
+events to run, or not.  Triggers must give a consistent result over
```
Comment (knelli2) [inline docs/Tutorials/EventsAndTriggers.md]: [nitpick] "At each check a trigger may fire, causing its associated events to run or not."

**Review by knelli2 — COMMENTED** [review]
Just a few nitpicks which you can choose to change or not.

Other than that I'm happy with this :)

**Comment by yoonso0-0** [comment]
Fixup LGTM :)

**Comment by wthrowe** [comment]
Squashed, including the grammar nits.

**Review by knelli2 — APPROVED** [review]

**Review by yoonso0-0 — APPROVED** [review]

**Comment by nilsdeppe** [comment]
Does this fix #1452?

**Comment by knelli2** [comment]
Partially? See one of my previous comments about it and @wthrowe's response.

Basically I think this is a good user tutorial, giving enough background info on events and triggers that users can modify input files without just copy/pasting blindly. However, if this were a dev guide, I'd be very lost on how everything actually works internally, how to write new ones (without just copy/pasting), and how in general these fit into the larger framework (corresponding actions and what they do).

**Review by kidder — APPROVED** [review]
